runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -21,64 +22,65 @@
-                workflowId, server, response.isActive());
-            return response.isActive();
-        });
+    void waitForWorkflow(UUID workflow, ManagementClient client) throws NetworkException,
+            TimeoutException {
+        long tries = runtime.getParameters().getWorkflowTimeout().getSeconds() /
+                runtime.getParameters().getWorkflowRetryRate().getSeconds();
+        for (long x = 0; x < tries; x++) {
+            if (!client.queryRequest(workflow).isActive()) {
+                return;
+            }
+            Sleep.sleepUninterruptibly(runtime.getParameters().getWorkflowRetryRate());
+            log.info("waitForWorkflow: waiting for {} on try {}", workflow, x);
+        }
+        log.debug("waitForWorkflow: workflow {} timeout", workflow);
+        throw new TimeoutException();
     }
 
     /**
      * Add a new node to the existing cluster.
      *
-     * @param endpoint Endpoint of the new node to be added to the cluster.
-     * @return True if completed successfully.
+     * @param endpointToAdd Endpoint of the new node to be added to the cluster.
+     * @throws WorkflowException throws a workflow exception on failure
      */
-    public boolean addNode(String endpoint) {
-        return layoutHelper(l -> {
+    public void addNode(String endpointToAdd) {
 
-            // Choosing the tail log server to run the workflow to optimize bulk reads on the
-            // same node.
-            List<String> logServers = l.getSegments().get(0).getStripes().get(0).getLogServers();
-            String server = logServers.get(logServers.size() - 1);
-            ManagementClient client = runtime.getRouter(server).getClient(ManagementClient.class);
-
-            // Dispatch add node request.
-            CreateWorkflowResponse addNodeResponse = client.addNodeRequest(endpoint);
+        for (int x = 0; x < runtime.getParameters().getWorkflowRetry(); x++) {
+            runtime.invalidateLayout();
+            Layout layout = runtime.getLayoutView().getLayout();
+            List<String> logServers = layout.getSegments().get(0).getStripes().get(0).getLogServers();
+            String orchestratorEndpoint = logServers.get(logServers.size() - 1);
+            ManagementClient client = runtime.getRouter(orchestratorEndpoint)
+                    .getClient(ManagementClient.class);
 
-            UUID workflowId = addNodeResponse.getWorkflowId();
 
-            while (isWorkflowActive(server, workflowId)) {
-                Sleep.MILLISECONDS.sleepUninterruptibly(layoutRefreshTimeout);
-            }
-
-            runtime.invalidateLayout();
+            try {
+                CreateWorkflowResponse resp = client.addNodeRequest(endpointToAdd);
+                log.info("addNode: Workflow id {} for {}", resp.workflowId, endpointToAdd);
+                waitForWorkflow(resp.getWorkflowId(), client);
+                Layout newLayout = null;
+                for (int y = 0; y < 3; y++) {
```

Comment:
hard-coded constant

Reply:
Ill just put this in the corfu runtime parameters. invalidating and check the layout a couple of times is something that we need to do in a couple of places.

Reply:
No. Retries is instance specific. Sometimes I might want to not retry and sometimes I might want to retry forever, on two parallel calls.

Make it a function parameter.

Reply:
I'd contrast this with the global rpc timeout. While it is  true that some functions may want a shorter or longer RPC time out, 99% of the functions are agnostic of the RPC timeout.

Reply:
This is a specific case, the invalidateRetry is only used where a layout is expected to change (i.e. currently attemptConsensus and after a workflow finishes executing), it won't be used for all layout invalidation. In that case, it seems a bit contrived that attempt consensus wants to invalidate the layout 3 times, but addnode wants to do it for for 5 times while both code paths are trying to achieve the exact thing. 

Let me know if you have a particular use case, otherwise I'll leave it as is, its a small change anyways.